src/core/control/xojfile/XmlParser.cpp

```diff
@@ -0,0 +1,814 @@
+                case TagType::STROKE:
+                    this->handler->finalizeStroke();
+                    break;
+                case TagType::TEXT:
+                    this->handler->finalizeText();
+                    break;
+                case TagType::IMAGE:
+                    this->handler->finalizeImage();
+                    break;
+                case TagType::TEXIMAGE:
+                    this->handler->finalizeTexImage();
+                    break;
+                default:
+                    break;
+            }
+            closeTag(currentTagType());
+            return xmlTextReaderRead(this->reader);
+        }
+        default:
+            g_warning("XML parser: Ignoring unexpected node type %d in layer", nodeType);
+            return xmlTextReaderRead(this->reader);
+    }
+}
+
+auto XmlParser::processAttachment() -> int {
+    xoj_assert(!this->hierarchy.empty());
+
+    const int nodeType = xmlTextReaderNodeType(this->reader);
+    switch (nodeType) {
+        case XML_ELEMENT_NODE: {
+            xoj_assert(this->hierarchy.top() == TagType::IMAGE || this->hierarchy.top() == TagType::TEXIMAGE ||
+                       this->hierarchy.top() == TagType::UNKNOWN);
+
+            const TagType tagType = openTag();
+
+            switch (tagType) {
+                case TagType::ATTACHMENT:
+                    parseAttachment();
+                    break;
+                default:
+                    g_warning("XML parser: Ignoring unexpected tag in image or TEX image: \"%s\"", currentName());
+                    break;
+            }
+            return xmlTextReaderRead(this->reader);
+        }
+        case XML_ELEMENT_DECL:
+            closeTag(currentTagType());
+            return xmlTextReaderRead(this->reader);
+        default:
+            g_warning("XML parser: Ignoring unexpected node type %d in image or TEX image", nodeType);
+            return xmlTextReaderRead(this->reader);
+    }
+}
+
+
+void XmlParser::parseXournalTag() {
+    const auto attributeMap = getAttributeMap();
+
+    std::string creator;
+    const auto optCreator = XmlParserHelper::getAttrib<std::string>("creator", attributeMap);
+    if (optCreator) {
+        creator = *optCreator;
+    } else {
+        // Compatibility: the creator attribute exists since 7017b71. Before that, only a version string was written
+        const auto optVersion = XmlParserHelper::getAttrib<std::string>("version", attributeMap);
+        if (optVersion) {
+            creator = "Xournal " + *optVersion;
+        } else {
+            creator = "Unknown";
+        }
+    }
+
+    const auto fileversion = XmlParserHelper::getAttribMandatory<int>("fileversion", attributeMap, 1);
+
+    this->handler->addXournal(creator, fileversion);
+}
+
+void XmlParser::parseMrWriterTag() {
+    const auto attributeMap = getAttributeMap();
+
+    std::string creator;
+    auto optVersion = XmlParserHelper::getAttrib<std::string>("verison", attributeMap);
```

Comment:
Ok, I've put all the attribute names in a file `XmlNames.h` and a namespace `XmlAttrs`. Two questions for you:

1. Should I make it more structured, that is, define `XmlAttrs::Page::WIDTH` and `XmlAttrs::Stroke::WIDTH` instead of just `XmlAttrs::WIDTH`?
2. I should probably do the same for tag names right?